### ISSUE TITLE
CLI to allow API calls

### ISSUE THREAD

**Issue by rmcrackan** [rmcrackan/Libation#683]
Since CLI already has auth tokens, this could be a good place to allow power users to make direct API calls.

**Comment by Mbucari** [comment]
One thing that will make this slightly difficult is that most of the types returned by AudibleApi don't have the source json. Currently, only classes that inherit from [DtoBase](https://github.com/rmcrackan/AudibleApi/blob/master/AudibleApi.Common/_DtoBase.cs) have it. We could just convert back to json, but that has the undesirable side effect of leaving a whole bunch of null properties in the converted json that were absent in the source json.

**Comment by wtanksleyjr** [comment]
> I assumed we’d just return the raw json. This is for power users anyway for sh...tuff we don’t want to fully support in Libation.

You have indeed discovered my kink.

Would it be possible to someday also have Hangover functionality in the CLI?

**Comment by rmcrackan** [comment]
Example?

**Comment by Mbucari** [comment]
> I assumed we’d just return the raw json.

Yeah, and I'm saying we need to modify AudibleApi to be able to do that.

**Comment by rmcrackan** [comment]
If you mean db queries: they do not belong in the CLI. I emphatically believe that a database is not an api, it’s an implementation detail and should always be open to change by the developers. For this reason, it’s in the debug tool, not on the CLI.

That said, nothing’s stopping you from using a 3rd party sqlite app. I highly highly advise against editing anything this way though.

**Comment by rmcrackan** [comment]
> Yeah, and I'm saying we need to modify AudibleApi to be able to do that.

Ah, now I understand

**Comment by wtanksleyjr** [comment]
It's never occurred to me to claim a database is an API. I was talking
about Libation as a whole, not just the new API feature. I apologize for
bringing this up in the wrong thread, because I can see why you would think
I meant that.

What I mean is that I would like to have more of Libation's existing
functionality from my headless server, without having to clumsily copy
parts of the config back and forth to my Windows laptop. But I'll ask for
that later in a different enhancement request, AFTER this wonderful new
functionality is merged. Again, I apologize for speaking out of turn on the
wrong thread.

-Wm

**Comment by rmcrackan** [comment]
@wtanksleyjr No worries. I was also replying on my phone earlier. I can type better and more thoroughly here from a real keyboard.

Hangover functionality is currently a box to interact with the raw database and the "deleted books" tab which is kinda a recycling bin for deleted books. The 2nd one doesn't sound practical for the CLI so I assumed you wanted the db functionality in the CLI -- so that's what I was addressing.

I agree that your request belongs in another ticket. I'm curious though of what specifically you had in mind. I don't use the CLI so I'm sure there are all kinds of blind-spots that haven't occurred to me. RE CLI: what are your pain points?

**Comment by wtanksleyjr** [comment]
Cool question. Let me answer re: my CLI pain points by mentioning my
greatest happy point: that the DEB installer puts in a place for the
database and config files, removing the need to guess how to manually
configure that. I'm hoping I can eventually figure out a similar way to
handle a docker-compose install or something like that.

Prior to that, my greatest pain point was how hard it was to get the CLI
running; now it's easier. It might be crazy to want to add accounts through
the CLI, but I'm also crazy that way ;) . You know, using the old "go to
this URL, log in, and then paste the new URL into here".

Generally speaking, I'd like to be able to fix the same kinds of problems
that one can fix with the GUI -- mark things as downloaded/not, remove
records that didn't work well, and so on. That's why I mentioned Hangover.

But it's enough to work on the main path: install, add accounts, scan, dump
info, liberate.

-Wm

**Comment by Mbucari** [comment]
Leaving this here is an example and for later reference.

https://api.audible.com/1.0/catalog/products?asins=B09LZ58X8G&response_groups=contributors,media,product_attrs,product_desc,product_extended_attrs,product_plan_details,product_plans,rating,review_attrs,reviews,sample,sku,relationships,category_ladders,claim_code_url,price,provided_review,series

**Comment by rmcrackan** [comment]
This is one of a number of far-reaching features that have been given a lot of thought for a long time with no commitments. I think it's fair to say we're not intending to do it in the near term.
